### PR TITLE
fixes and refactoring of s3_bucket

### LINKED ISSUE #15603: Faulty parsing of JSON template/file lookups (aka JSON double-quotes to single-quotes conversion problem)
##### ISSUE TYPE
- Bug Report
##### ANSIBLE VERSION

```
2.1.0
```
##### CONFIGURATION

Default from Git master
##### OS / ENVIRONMENT

Debian Jessie x86_64
##### SUMMARY

JSON `lookup('template', foobar)` content gets returned as variable that gets reparsed as dict (afaik), which somewhere gets serialized to string before it gets to the module and converts valid-JSON double-quotes to single-quotes so services or modules that expect valid JSON will fail with such data.

The problem is that Jinja2 templating is done recursively and I could not find any combination of `convert_*` arguments, Jinja2 filters, nor `#jinja2` overrides that would return the correct JSON from the template to the module and the classic override of putting a space before the string start isn't allowed when contacting AWS services. I believe the same issue exists with `lookup('file')` and any other lookup that would return JSON.

<br />

Problem code:
- https://github.com/ansible/ansible/blob/devel/lib/ansible/template/__init__.py#L333-L335
  - Returned JSON from template ends up being templated again since it starts with `{` and returns a dict.

<br/>
- Bug fix PR that was fixing the symptom in `iam_policy` module with `policy_json` arg: https://github.com/ansible/ansible-modules-core/pull/3019
- Another new (unreported yet) bug caused by this logic breaks `policy` argument in `s3_bucket` module here if you use a lookup: https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/amazon/s3_bucket.py#L211

I'd probably fix this myself but I just don't have the time to do it :(
##### STEPS TO REPRODUCE

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "AddPerm",
      "Effect": "Allow",
      "Principal": "*",
      "Action": "s3:GetObject",
      "Resource": "arn:aws:s3:::{{ item }}"
    }
  ]
}
```

```

---

- hosts: all
  connection: local
  gather_facts: False

  tasks:
    - name: Setup S3 Bucket
      s3_bucket:
        name: "{{ item }}"
        region: "us-west-2"
        policy: "{{ lookup('template','./bucket-public.json.j2', convert_data=False) }}"
        with_item:
          - some_bucket_name
```
##### EXPECTED RESULTS

JSON posted to AWS is valid and bucket is created
##### ACTUAL RESULTS

```
PLAYBOOK: deploy.yml ***********************************************************
1 plays in deploy.yml

PLAY [all] *********************************************************************

TASK [Setup S3 Bucket] *********************************************************
task path: /home/username/checkout/git-repo-name/deploy.yml:12
File lookup using /home/username/checkout/git-repo-name/bucket-public.json.j2 as file
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: username
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1461682432.3-159666311309506 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1461682432.3-159666311309506 `" )'
<localhost> PUT /tmp/tmph4kIsC TO /home/username/.ansible/tmp/ansible-tmp-1461682432.3-159666311309506/s3_bucket
<localhost> EXEC /bin/sh -c 'LANG=C LC_ALL=C LC_MESSAGES=C /usr/bin/python /home/username/.ansible/tmp/ansible-tmp-1461682432.3-159666311309506/s3_bucket; rm -rf "/home/username/.ansible/tmp/ansible-tmp-1461682432.3-159666311309506/" > /dev/null 2>&1'
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"aws_access_key": null, "aws_secret_key": null, "ec2_url": null, "force": false, "name": "some_bucket_name", "policy": "{'Version': '2012-10-17', 'Statement': [{'Action': 's3:GetObject', 'Principal': '*', 'Resource': 'arn:aws:s3:::some_bucket_name', 'Effect': 'Allow', 'Sid': 'AddPerm'}]}", "profile": null, "region": "us-west-2", "requester_pays": false, "s3_url": null, "security_token": null, "state": "present", "tags": {}, "validate_certs": true, "versioning": false}, "module_name": "s3_bucket"}, "msg": "Policies must be valid JSON and the first byte must be '{'"}


NO MORE HOSTS LEFT *************************************************************
    to retry, use: --limit @deploy.retry

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1
```

### PR DESCRIPTION
**Pull request by bcoca**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

s3_bucket
##### ANSIBLE VERSION

```
2.2
```
##### SUMMARY

policy should now accept and handle correctly both data structures or JSON strings
other minor fixes, removed unused vars, set correct type in args spec, clarified func name, etc
refactored code to make conditions clearer

fixes https://github.com/ansible/ansible/issues/15603

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @bcoca for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by wimnat** [comment]
@bcoca as cleanup can you also remove the import XML statement please?

File: cloud/amazon/s3_bucket.py [hunk 1]
```diff
@@ -146,73 +143,57 @@ def create_bucket(connection, module, location):
...
-        if json.loads(current_policy) != json.loads(policy):
+
+    if policy is not None:
+        # Deal with policy if either JSON formatted string or just data structure
+        if isinstance(policy, basestring):
+            compare_policy = json.dumps(policy)
+            load_policy = policy
+        else:
+            compare_policy = policy
+            load_policy = json.loads(policy)
```
Comment (abadger) [inline cloud/amazon/s3_bucket.py]: This is reversed (Note I also threw in strip() to take care of the issue reported on the mailing list):

``` python
if isinstance(policy, basestring):
    compare_policy = json.loads(policy)
    load_policy = policy.strip()
else:
    compare_policy = policy
    load_policy = json.dumps(policy)
```

**Comment by abadger** [comment]
refactoring looks good except for the one spot I noted.  You could instead use the 'jsonarg' type in the argspec and then that snippet of code will reduce to unconditionally doing:

``` python
compare_policy = json.loads(policy)
```

**Comment by abadger** [comment]
@wimnat It looks like `import xml.etree.ElementTree as ET` is used later to parse the string in `get_request_payment_status()`.  How were you thinking that should be removed?

**Comment by gregdek** [comment]
@bcoca A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@bcoca Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@bcoca A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/amazon/s3_bucket.py [hunk 2]
```diff
@@ -146,73 +143,57 @@ def create_bucket(connection, module, location):
...
+
+        if current_policy is None or json.loads(current_policy) != compare_policy:
             try:
-                bucket.set_policy(policy)
+                bucket.set_policy(load_policy)
```
Comment (bretmartin) [inline cloud/amazon/s3_bucket.py]: `UnboundLocalError: local variable 'load_policy' referenced before assignment`

**Comment by gregdek** [comment]
@bcoca Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@bcoca A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by borgstrom** [comment]
It is currently **impossible** to manage S3 buckets with policies as it stands, which are required if you want to write ELB logs into a bucket.

@bcoca @wimnat what can be done to get this merged so we can at least just build our own version of Ansible until a new release is officially cut?

**Comment by bcoca** [comment]
updated to account for new code, not sure how to make ceph part deal with updates so left TODO.

please test as I do NOT have access to aws account currently.

**Comment by borgstrom** [comment]
Thanks for looking at this, @bcoca!

Was `type: json` added after 2.1 was released?  I ran my playbook with keep remote files on, exploded the s3_bucket step and then replaced the module with the version from this PR and I got:

```
#2556 ❯❯❯ python ~/.ansible/tmp/ansible-tmp-1469549941.57-172683670523748/s3_bucket excommunicate

{"msg": "implementation error: unknown type json requested for policy", "failed": true, "invocation": {"module_args": {"force": false, "name": "us-east-1-devops-facet-elb-logs", "tags": {"cluster": "devops", "pod": "devops"}, "region": "us-east-1", "ceph": false, "state": "present", "policy": {"Version": "2012-10-17", "Id": "Policy1469475720170", "Statement": [{"Action": ["s3:PutObject"], "Principal": {"AWS": ["127311923021"]}, "Resource": "arn:aws:s3:::us-east-1-devops-facet-elb-logs/*", "Effect": "Allow", "Sid": "Stmt1469475716322"}]}, "requester_pays": "no", "validate_certs": true, "versioning": "no"}}}

#2557 ❯❯❯ ansible --version
ansible 2.1.0.0
  config file = 
  configured module search path = Default w/o overrides
```

**Comment by gregdek** [comment]
Thanks @bcoca. To the current maintainers, @wimnat please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by borgstrom** [comment]
If I change type to `str` for policy, then I get a different error:

```
Traceback (most recent call last):
  File "/Users/evanborgstrom/.ansible/tmp/ansible-tmp-1469550477.6-214270110872560/s3_bucket", line 215, in <module>
    exitcode = debug(sys.argv[1], zipped_mod, ZIPLOADER_PARAMS)
  File "/Users/evanborgstrom/.ansible/tmp/ansible-tmp-1469550477.6-214270110872560/s3_bucket", line 188, in debug
    main()
  File "/Users/evanborgstrom/.ansible/tmp/ansible-tmp-1469550477.6-214270110872560/debug_dir/ansible_module_s3_bucket.py", line 424, in main
    create_or_update_bucket(connection, module, location)
  File "/Users/evanborgstrom/.ansible/tmp/ansible-tmp-1469550477.6-214270110872560/debug_dir/ansible_module_s3_bucket.py", line 312, in create_or_update_bucket
    _create_or_update_bucket(connection, module, location)
  File "/Users/evanborgstrom/.ansible/tmp/ansible-tmp-1469550477.6-214270110872560/debug_dir/ansible_module_s3_bucket.py", line 206, in _create_or_update_bucket
    bucket.set_policy(load_policy)
NameError: global name 'load_policy' is not defined
```

On line 206 `load_policy` should just be `policy`

After that change it works!

```
#2583 ❯❯❯ AWS_PROFILE=nwdev python /Users/evanborgstrom/.ansible/tmp/ansible-tmp-1469550477.6-214270110872560/s3_bucket excommunicate

{"name": "us-east-1-devops-facet-elb-logs", "tags": {"cluster": "devops", "pod": "devops"}, "changed": true, ... }
```

**Comment by borgstrom** [comment]
In favor of fixing 2.1 I propose that this change be merged with type set to `str` and the documentation updated so that it shows the way to load a policy is:

```
policy: "{{ lookup('file','policy.json')|to_json }}"
```

Then in 2.2 type can be changed to `json`, which IIUC, will allow the policy to be attached with or without the `|to_json` filter.

@bcoca @wimnat -- what do you think?

**Comment by ryansb** [comment]
I'd be strongly against changing the argument type later. The JSON type exists now, and I'd rather use that and have this refactor go in with 2.2 than try to backport.

**Comment by borgstrom** [comment]
@ryansb - Can you explain why you'd be against it?

By accepting the json type as the "fix" it leaves all of us who prefer to track stable in a position where the s3_bucket module is broken until 2.2 is officially released.

**Comment by ryansb** [comment]
The modules & core are released as one tarball, so the use of `json` wouldn't be released until 2.2 is considered stable. 

I'm against it because people who wrote playbooks using the `str` definition `{{ lookup('file','policy.json')|to_json }}` would have their playbooks broken by switching to JSON, and we don't want to break playbooks.

**Comment by borgstrom** [comment]
Sure, I can accept that, except they're broken right now as-is in 2.1...  So anyone that you're worried about breaking is already broken.

I also definitely want to use the `json` type in the long run, but for now I just want a fix that can easily applied to the 2.1-stable branch so I can stop printing messages telling people that they need to create buckets and install policies by hand because I need features in 2.1 (thus can't roll back to 2.0.2) and don't want to start using the dev branches.

**Comment by bcoca** [comment]
I had code to deal with the argument w/o using type='json', I'll try to backport this fix with that code once it is confirmed this works. 

The backport would be part of a 'possible' 2.1.2 release.

File: cloud/amazon/s3_bucket.py [hunk 3]
```diff
@@ -160,73 +158,58 @@ def _create_bucket(connection, module, location):
...
-            
-        if json.loads(current_policy) != json.loads(policy):
+
+    if policy is not None:
+        compare_policy = json.loads(policy)
+
+        if current_policy is None or json.loads(current_policy) != compare_policy:
             try:
-                bucket.set_policy(policy)
+                bucket.set_policy(load_policy)
```
Comment (ryansb) [inline cloud/amazon/s3_bucket.py]: Is this supposed to just be `policy`? When testing this fails for me with:

```
Traceback (most recent call last):
  File "/tmp/ryansb/ansible_I4fJea/ansible_module_s3_bucket.py", line 429, in <module>
    main()
  File "/tmp/ryansb/ansible_I4fJea/ansible_module_s3_bucket.py", line 424, in main
    create_or_update_bucket(connection, module, location)
  File "/tmp/ryansb/ansible_I4fJea/ansible_module_s3_bucket.py", line 312, in create_or_update_bucket
    _create_or_update_bucket(connection, module, location)
  File "/tmp/ryansb/ansible_I4fJea/ansible_module_s3_bucket.py", line 206, in _create_or_update_bucket
    bucket.set_policy(load_policy)
NameError: global name 'load_policy' is not defined
```

Comment (bcoca) [inline cloud/amazon/s3_bucket.py]: fixing it, was typo

**Comment by ryansb** [comment]
This works for me now, testing with:

```
- name: Make an S3 bucket with a broken policy
    s3_bucket:
      region: us-east-1
      name: ryansb-test-ansible-module
      policy: " {{ lookup('file', './2016-07-s3-bucket-policy.json') }}"
```

**Comment by ryansb** [comment]
Test failure on Travis is due to a bad `version_added` on a module: fixed https://github.com/ansible/ansible-modules-extras/commit/a69aa6008160014b902a075be12c7abbf19a0d3c

**Comment by ryansb** [comment]
@bcoca will push a backwards compatible version to `stable` separately.

**Comment by isopy** [comment]
I'm still seeing an an error when using templates with the s3_bucket.py from this PR:

**Playbook task:**

```
s3_bucket:
  name: "{{ inventory_hostname }}"
  region: "{{ aws_region }}"
  policy: "{{ lookup('template', playbook_dir + '/roles/aws_provision_s3_bucket/templates/default_policy.json.j2') }}"
  state: present
```

**Fails with this error:**
FAILED! => {"changed": false, "failed": true, "msg": "implementation error: unknown type json requested for policy"}

**default_policy.json.j2:**

```
{
        "Version": "2008-10-17",
        "Statement": [
                {
                        "Sid": "",
                        "Effect": "Allow",
                        "Principal": { 
                                "AWS": "arn:aws:iam::{{ aws_account_number }}:user/{{ created_iam_user.user_name }}"
                        },
                        "Action": [
                                "s3:ListBucket",
                                "s3:GetBucketLocation",
                                "s3:ListBucketMultipartUploads"
                        ],
                        "Resource": "arn:aws:s3:::{{ inventory_hostname }}/*"
                },
                {
                        "Sid": "{{ inventory_hostname }} Update",
                        "Effect": "Allow",
                        "Principal": { 
                                "AWS": "arn:aws:iam::{{ aws_account_number }}:user/{{ created_iam_user.user_name }}"
                        },
                        "Action": [
                                "s3:PutObject",
                                "s3:AbortMultipartUpload",
                                "s3:GetObject*",
                                "s3:DeleteObject",
                                "s3:ListMultipartUploadParts",
                                "s3:List*"
                        ],
                        "Resource": "arn:aws:s3:::{{ inventory_hostname }}/*"
                }
        ]
}

```

**Comment by ryansb** [comment]
What version of ansible are you using the `s3_bucket` module from this PR?

**Comment by isopy** [comment]
Apologies for not including that info. I'm on 2.1.0.0

**Comment by ryansb** [comment]
@michaeljfoxy This patch won't work there - it requires a `json` type introduced in the 2.2 development cycle.

**Comment by isopy** [comment]
Ah gotcha. Any ETA on a 2.2 stable branch or suggestions how I can grab it now? Getting S3 provisioned is a major milestone in my current project so I'm itching to test it out.

**Comment by clott-memberclicks** [comment]
This is busted for us as well. S3 provisioning just isn't feasible without some workable way to apply a policy.

**Pull request merged**